File: crates/adapters/thetadata/src/decode.rs
// -------------------------------------------------------------------------------------------------
//  Copyright (C) 2015-2026 Nautech Systems Pty Ltd. All rights reserved.
//  https://nautechsystems.io
//
//  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
//  You may not use this file except in compliance with the License.
//  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// -------------------------------------------------------------------------------------------------

//! Wire DTO → Nautilus domain type decoders.
//!
//! Two timestamp conventions to bridge — both treat the source as **America/New_York wall clock**
//! and convert to UTC `UnixNanos` via DST-aware `chrono-tz`:
//!
//! - **REST** rows carry ISO 8601 millisecond strings (no offset), e.g.
//!   `"2023-12-19T09:30:22.025"`.
//! - **WebSocket** frames split the timestamp into `date` (`YYYYMMDD` int) and `ms_of_day`
//!   (milliseconds since 00:00:00 ET).
//!
//! Ambiguous local times during the fall-back DST transition resolve to the earlier UTC
//! instant (standard time); non-existent times during the spring-forward transition are
//! rejected.

use anyhow::{Context, Result, anyhow, bail};
use chrono::{LocalResult, NaiveDate, NaiveDateTime, NaiveTime, TimeZone};
use chrono_tz::{America::New_York, Tz};
use nautilus_core::UnixNanos;
use nautilus_model::{
    data::{Bar, BarType, QuoteTick, TradeTick},
    enums::AggressorSide,
    identifiers::{InstrumentId, TradeId},
    types::{Price, Quantity},
};

use crate::{
    enums::aggressor_from_condition,
    types::{RestOhlcRow, RestQuoteRow, RestTradeRow, WsQuoteFrame, WsTradeFrame},
};

const MS_TO_NS: u64 = 1_000_000;
const ET_TZ: Tz = New_York;

/// Combines a `YYYYMMDD` date integer and `ms_of_day` (Eastern Time) into a `UnixNanos`.
///
/// # Errors
///
/// Returns an error if the date integer is not a valid calendar date, `ms_of_day` is out of
/// range for a single day (86 400 000 ms), or the local time does not exist (spring-forward gap).
pub fn ws_timestamp_to_unix_nanos(date: u32, ms_of_day: u64) -> Result<UnixNanos> {
    if ms_of_day >= 86_400_000 {
        bail!("ms_of_day out of range: {ms_of_day}");
    }
    let year = (date / 10_000) as i32;
    let month = (date / 100) % 100;
    let day = date % 100;
    let naive_date = NaiveDate::from_ymd_opt(year, month, day)
        .ok_or_else(|| anyhow!("invalid date: {date}"))?;
    let hours = (ms_of_day / 3_600_000) as u32;
    let mins = ((ms_of_day % 3_600_000) / 60_000) as u32;
    let secs = ((ms_of_day % 60_000) / 1_000) as u32;
    let millis = (ms_of_day % 1_000) as u32;
    let naive_time = NaiveTime::from_hms_milli_opt(hours, mins, secs, millis)
        .ok_or_else(|| anyhow!("invalid time-of-day: {ms_of_day}"))?;
    et_naive_to_unix_nanos(NaiveDateTime::new(naive_date, naive_time))
}

/// Parses an ISO 8601 millisecond timestamp returned by the REST API and converts to `UnixNanos`.
///
/// Accepts `"YYYY-MM-DDTHH:MM:SS.SSS"` (no offset). The value is interpreted as
/// America/New_York wall clock.
///
/// # Errors
///
/// Returns an error if the string cannot be parsed or the local time does not exist.
pub fn parse_rest_timestamp(s: &str) -> Result<UnixNanos> {
    let naive = NaiveDateTime::parse_from_str(s, "%Y-%m-%dT%H:%M:%S%.3f")
        .with_context(|| format!("invalid REST timestamp {s:?}"))?;
    et_naive_to_unix_nanos(naive)
}

/// Converts a naive ET wall-clock datetime to `UnixNanos`.
///
/// # Errors
///
/// Returns an error if the local time does not exist (DST spring-forward gap). Ambiguous times
/// (fall-back overlap) resolve to the earlier UTC instant.
pub fn et_naive_to_unix_nanos(naive: NaiveDateTime) -> Result<UnixNanos> {
    let resolved = match ET_TZ.from_local_datetime(&naive) {
        LocalResult::Single(dt) => dt,
        LocalResult::Ambiguous(earlier, _) => earlier,
        LocalResult::None => bail!("local time {naive} does not exist in America/New_York"),
    };
    let millis = resolved.timestamp_millis();
    if millis < 0 {
        bail!("timestamp pre-epoch");
    }
    Ok(UnixNanos::from((millis as u64) * MS_TO_NS))
}

impl RestQuoteRow {
    /// Decodes this row into a Nautilus [`QuoteTick`].
    ///
    /// `price_precision` and `size_precision` come from the instrument metadata (typically
    /// `(2, 0)` for US options).
    ///
    /// # Errors
    ///
    /// Returns an error if the row's timestamp cannot be parsed.
    pub fn to_quote_tick(
        &self,
        instrument_id: InstrumentId,
        price_precision: u8,
        size_precision: u8,
        ts_init: UnixNanos,
    ) -> Result<QuoteTick> {
        let ts_event = parse_rest_timestamp(&self.timestamp)?;
        let bid_price = Price::new(self.bid, price_precision);
        let ask_price = Price::new(self.ask, price_precision);
        let bid_size = Quantity::new(f64::from(self.bid_size), size_precision);
        let ask_size = Quantity::new(f64::from(self.ask_size), size_precision);
        QuoteTick::new_checked(
            instrument_id,
            bid_price,
            ask_price,
            bid_size,
            ask_size,
            ts_event,
            ts_init,
        )
        .map_err(Into::into)
    }
}

impl RestTradeRow {
    /// Decodes this row into a Nautilus [`TradeTick`].
    ///
    /// The aggressor side is derived from the trade condition code (codes 145/146 only).
    ///
    /// # Errors
    ///
    /// Returns an error if the row's timestamp cannot be parsed or the size is non-positive.
    pub fn to_trade_tick(
        &self,
        instrument_id: InstrumentId,
        price_precision: u8,
        size_precision: u8,
        ts_init: UnixNanos,
    ) -> Result<TradeTick> {
        let ts_event = parse_rest_timestamp(&self.timestamp)?;
        let price = Price::new(self.price, price_precision);
        let size = Quantity::new(self.size as f64, size_precision);
        let aggressor: AggressorSide = aggressor_from_condition(self.condition);
        let trade_id = TradeId::new(self.sequence.to_string().as_str());
        TradeTick::new_checked(instrument_id, price, size, aggressor, trade_id, ts_event, ts_init)
            .map_err(Into::into)
    }
}

impl WsQuoteFrame {
    /// Decodes this inbound WebSocket quote into a Nautilus [`QuoteTick`].
    ///
    /// # Errors
    ///
    /// Returns an error if the embedded `date + ms_of_day` cannot be combined into a valid
    /// timestamp.
    pub fn to_quote_tick(
        &self,
        instrument_id: InstrumentId,
        price_precision: u8,
        size_precision: u8,
        ts_init: UnixNanos,
    ) -> Result<QuoteTick> {
        let ts_event = ws_timestamp_to_unix_nanos(self.quote.date, self.quote.ms_of_day)?;
        let bid_price = Price::new(self.quote.bid, price_precision);
        let ask_price = Price::new(self.quote.ask, price_precision);
        let bid_size = Quantity::new(f64::from(self.quote.bid_size), size_precision);
        let ask_size = Quantity::new(f64::from(self.quote.ask_size), size_precision);
        QuoteTick::new_checked(
            instrument_id,
            bid_price,
            ask_price,
            bid_size,
            ask_size,
            ts_event,
            ts_init,
        )
        .map_err(Into::into)
    }
}

impl WsTradeFrame {
    /// Decodes this inbound WebSocket trade into a Nautilus [`TradeTick`].
    ///
    /// Aggressor side derives from the trade condition code (codes 145/146 only).
    ///
    /// # Errors
    ///
    /// Returns an error if the timestamp cannot be combined or the size is non-positive.
    pub fn to_trade_tick(
        &self,
        instrument_id: InstrumentId,
        price_precision: u8,
        size_precision: u8,
        ts_init: UnixNanos,
    ) -> Result<TradeTick> {
        let ts_event = ws_timestamp_to_unix_nanos(self.trade.date, self.trade.ms_of_day)?;
        let price = Price::new(self.trade.price, price_precision);
        let size = Quantity::new(self.trade.size as f64, size_precision);
        let aggressor = aggressor_from_condition(self.trade.condition);
        let trade_id = TradeId::new(self.trade.sequence.to_string().as_str());
        TradeTick::new_checked(instrument_id, price, size, aggressor, trade_id, ts_event, ts_init)
            .map_err(Into::into)
    }
}

impl RestOhlcRow {
    /// Decodes this row into a Nautilus [`Bar`].
    ///
    /// `bar_type` carries the instrument id, aggregation, and price-type — construct it at the
    /// call site from the historical request parameters.
    ///
    /// # Errors
    ///
    /// Returns an error if the row's timestamp cannot be parsed or OHLC invariants
    /// (`high >= low`, etc.) are violated.
    pub fn to_bar(
        &self,
        bar_type: BarType,
        price_precision: u8,
        size_precision: u8,
        ts_init: UnixNanos,
    ) -> Result<Bar> {
        let ts_event = parse_rest_timestamp(&self.timestamp)?;
        let open = Price::new(self.open, price_precision);
        let high = Price::new(self.high, price_precision);
        let low = Price::new(self.low, price_precision);
        let close = Price::new(self.close, price_precision);
        let volume = Quantity::new(self.volume as f64, size_precision);
        Bar::new_checked(bar_type, open, high, low, close, volume, ts_event, ts_init)
            .map_err(Into::into)
    }
}

#[cfg(test)]
mod tests {
    use nautilus_model::{
        data::BarType,
        enums::{BarAggregation, PriceType},
        identifiers::{InstrumentId, Symbol, Venue},
    };
    use rstest::*;
    use ustr::Ustr;

    use super::*;

    fn make_instrument_id() -> InstrumentId {
        InstrumentId::new(
            Symbol::new("AAPL240315C00180000"),
            Venue::new(Ustr::from("THETADATA")),
        )
    }

    #[rstest]
    fn test_ws_timestamp_combines_date_and_ms_in_et() {
        // 2023-12-19 09:30:22.025 ET (EST, UTC-5) → 2023-12-19 14:30:22.025 UTC
        // = 1_702_996_222_025 ms since epoch.
        let unix = ws_timestamp_to_unix_nanos(20_231_219, 34_222_025).unwrap();
        assert_eq!(u64::from(unix), 1_702_996_222_025_000_000);
    }

    #[rstest]
    fn test_ws_timestamp_rejects_out_of_range_ms() {
        assert!(ws_timestamp_to_unix_nanos(20_231_219, 86_400_000).is_err());
    }

    #[rstest]
    fn test_ws_timestamp_rejects_invalid_date() {
        assert!(ws_timestamp_to_unix_nanos(20_231_232, 0).is_err());
    }

    #[rstest]
    fn test_ws_timestamp_summer_uses_edt_offset() {
        // 2024-07-15 09:30:00 ET (EDT, UTC-4) → 13:30:00 UTC.
        let unix = ws_timestamp_to_unix_nanos(20_240_715, 34_200_000).unwrap();
        // Sanity check: should be 4 hours after midnight UTC + 9.5h.
        let expected_secs = chrono::NaiveDate::from_ymd_opt(2024, 7, 15)
            .unwrap()
            .and_hms_opt(13, 30, 0)
            .unwrap()
            .and_utc()
            .timestamp() as u64
            * 1_000_000_000;
        assert_eq!(u64::from(unix), expected_secs);
    }

    #[rstest]
    fn test_parse_rest_timestamp_basic() {
        let unix = parse_rest_timestamp("2023-12-19T09:30:22.025").unwrap();
        assert_eq!(u64::from(unix), 1_702_996_222_025_000_000);
    }

    #[rstest]
    fn test_parse_rest_timestamp_rejects_garbage() {
        assert!(parse_rest_timestamp("not-a-timestamp").is_err());
    }

    #[rstest]
    fn test_rest_quote_row_to_quote_tick() {
        let row = RestQuoteRow {
            symbol: "AAPL".into(),
            expiration: "2024-03-15".into(),
            strike: 180.0,
            right: "call".into(),
            timestamp: "2024-03-15T10:00:00.000".into(),
            bid_size: 5,
            bid_exchange: 10,
            bid: 1.23,
            bid_condition: 50,
            ask_size: 7,
            ask_exchange: 10,
            ask: 1.25,
            ask_condition: 50,
        };
        let tick = row.to_quote_tick(make_instrument_id(), 2, 0, UnixNanos::default()).unwrap();
        assert_eq!(tick.bid_price.to_string(), "1.23");
        assert_eq!(tick.ask_price.to_string(), "1.25");
        assert_eq!(tick.bid_size, Quantity::from(5));
        assert_eq!(tick.ask_size, Quantity::from(7));
    }

    #[rstest]
    #[case(0, AggressorSide::NoAggressor)]
    #[case(145, AggressorSide::Buyer)]
    #[case(146, AggressorSide::Seller)]
    fn test_rest_trade_row_to_trade_tick_aggressor_mapping(
        #[case] condition: u32,
        #[case] expected: AggressorSide,
    ) {
        let row = RestTradeRow {
            symbol: "AAPL".into(),
            expiration: "2024-03-15".into(),
            strike: 180.0,
            right: "call".into(),
            timestamp: "2024-03-15T10:00:00.000".into(),
            price: 1.24,
            size: 10,
            exchange: 10,
            condition,
            sequence: 42,
        };
        let tick = row.to_trade_tick(make_instrument_id(), 2, 0, UnixNanos::default()).unwrap();
        assert_eq!(tick.aggressor_side, expected);
        assert_eq!(tick.trade_id.to_string(), "42");
        assert_eq!(tick.size, Quantity::from(10));
    }

    #[rstest]
    fn test_rest_trade_row_rejects_zero_size() {
        let row = RestTradeRow {
            symbol: "AAPL".into(),
            expiration: "2024-03-15".into(),
            strike: 180.0,
            right: "call".into(),
            timestamp: "2024-03-15T10:00:00.000".into(),
            price: 1.24,
            size: 0,
            exchange: 10,
            condition: 0,
            sequence: 42,
        };
        assert!(row.to_trade_tick(make_instrument_id(), 2, 0, UnixNanos::default()).is_err());
    }

    #[rstest]
    fn test_rest_ohlc_row_to_bar() {
        let row = RestOhlcRow {
            symbol: "AAPL".into(),
            expiration: "2024-03-15".into(),
            strike: 180.0,
            right: "call".into(),
            timestamp: "2024-03-15T10:00:00.000".into(),
            open: 1.20,
            high: 1.30,
            low: 1.18,
            close: 1.25,
            volume: 1000,
            count: 50,
            vwap: 1.24,
        };
        let bar_type = BarType::new(
            make_instrument_id(),
            nautilus_model::data::BarSpecification::new(
                1,
                BarAggregation::Minute,
                PriceType::Last,
            ),
            nautilus_model::enums::AggregationSource::External,
        );
        let bar = row.to_bar(bar_type, 2, 0, UnixNanos::default()).unwrap();
        assert_eq!(bar.open.to_string(), "1.20");
        assert_eq!(bar.high.to_string(), "1.30");
        assert_eq!(bar.low.to_string(), "1.18");
        assert_eq!(bar.close.to_string(), "1.25");
        assert_eq!(bar.volume, Quantity::from(1000));
    }

    #[rstest]
    fn test_ws_quote_frame_to_quote_tick() {
        use crate::types::{WsContract, WsHeader, WsQuoteBody};
        let frame = WsQuoteFrame {
            header: WsHeader { status: "CONNECTED".into(), kind: "QUOTE".into() },
            contract: WsContract {
                security_type: "OPTION".into(),
                root: "SPXW".into(),
                expiration: 20_240_315,
                strike: 480_000,
                right: "C".into(),
            },
            quote: WsQuoteBody {
                ms_of_day: 34_222_025,
                bid_size: 7,
                bid_exchange: 5,
                bid: 110.2,
                bid_condition: 50,
                ask_size: 7,
                ask_exchange: 5,
                ask: 110.5,
                ask_condition: 50,
                date: 20_231_219,
            },
        };
        let tick = frame
            .to_quote_tick(make_instrument_id(), 2, 0, UnixNanos::default())
            .unwrap();
        assert_eq!(tick.bid_price.to_string(), "110.20");
        assert_eq!(tick.ask_price.to_string(), "110.50");
        assert_eq!(tick.bid_size, Quantity::from(7));
    }

    #[rstest]
    #[case(0, AggressorSide::NoAggressor)]
    #[case(145, AggressorSide::Buyer)]
    #[case(146, AggressorSide::Seller)]
    fn test_ws_trade_frame_to_trade_tick_aggressor(
        #[case] condition: u32,
        #[case] expected: AggressorSide,
    ) {
        use crate::types::{WsContract, WsHeader};
        let frame = WsTradeFrame {
            header: WsHeader { status: "CONNECTED".into(), kind: "TRADE".into() },
            contract: WsContract {
                security_type: "OPTION".into(),
                root: "AAPL".into(),
                expiration: 20_231_222,
                strike: 2_000_000,
                right: "C".into(),
            },
            trade: crate::types::WsTradeBody {
                ms_of_day: 34_389_945,
                sequence: 42,
                size: 10,
                condition,
                price: 0.31,
                exchange: 31,
                date: 20_231_219,
            },
        };
        let tick = frame
            .to_trade_tick(make_instrument_id(), 2, 0, UnixNanos::default())
            .unwrap();
        assert_eq!(tick.aggressor_side, expected);
        assert_eq!(tick.trade_id.to_string(), "42");
    }

    #[rstest]
    fn test_rest_ohlc_row_rejects_invariant_violation() {
        let row = RestOhlcRow {
            symbol: "AAPL".into(),
            expiration: "2024-03-15".into(),
            strike: 180.0,
            right: "call".into(),
            timestamp: "2024-03-15T10:00:00.000".into(),
            open: 1.20,
            high: 1.10, // < open — invalid
            low: 1.05,
            close: 1.15,
            volume: 100,
            count: 5,
            vwap: 1.12,
        };
        let bar_type = BarType::new(
            make_instrument_id(),
            nautilus_model::data::BarSpecification::new(1, BarAggregation::Minute, PriceType::Last),
            nautilus_model::enums::AggregationSource::External,
        );
        assert!(row.to_bar(bar_type, 2, 0, UnixNanos::default()).is_err());
    }
}

File: nautilus_trader/adapters/thetadata/decode.py
"""
ThetaData v3 wire-format decoders. Pure functions: REST/WS row → Nautilus type.

Wire format: docs/architecture/notes/thetadata-wire-format.md.

Timestamps: v3 carries ISO 8601 strings ("2024-06-20T09:30:00.334") in ET local
time, no timezone suffix. Converted to UTC nanos via zoneinfo. Spring-forward
gap times raise; fall-back ambiguity resolved by caller via `fold` (default 0).

Aggressor: only OPRA condition codes 145 (Buyer) and 146 (Seller) produce an
inferred aggressor; everything else → NO_AGGRESSOR. R3 fills downstream.
"""

from __future__ import annotations

from datetime import datetime
from zoneinfo import ZoneInfo

from msgspec import DecodeError

from nautilus_trader.model.data import Bar, BarType, QuoteTick, TradeTick
from nautilus_trader.model.enums import AggressorSide
from nautilus_trader.model.identifiers import InstrumentId, TradeId
from nautilus_trader.model.objects import Price, Quantity


_NY = ZoneInfo("America/New_York")

# Cache: (date_yyyymmdd, fold) -> UTC nanos at midnight ET on that date.
_DATE_CACHE: dict[tuple[int, int], int] = {}

_AGGRESSOR_BUYER = 145
_AGGRESSOR_SELLER = 146


def _cached_date_base(date_yyyymmdd: int, fold: int = 0) -> int:
    """UTC nanos at midnight ET on `date_yyyymmdd`, cached per (date, fold)."""
    key = (date_yyyymmdd, fold)
    base = _DATE_CACHE.get(key)
    if base is not None:
        return base
    y = date_yyyymmdd // 10000
    m = (date_yyyymmdd // 100) % 100
    d = date_yyyymmdd % 100
    base = int(datetime(y, m, d, 0, 0, 0, tzinfo=_NY, fold=fold).timestamp() * 1_000_000_000)
    _DATE_CACHE[key] = base
    return base


def date_ms_to_unix_nanos(date_yyyymmdd: int, ms_of_day: int, fold: int = 0) -> int:
    """Convert ET-local (date, ms_of_day) to UTC nanos.

    Kept for backwards compat with the date-base cache and DST tests; v3 wire
    format uses ISO timestamps (see iso_ts_to_unix_nanos). Both paths share
    `_DATE_CACHE` and produce identical UTC nanos for the same wall-clock.
    """
    if not (0 <= ms_of_day < 86_400_000):
        raise DecodeError(f"ms_of_day out of range: {ms_of_day}")

    y = date_yyyymmdd // 10000
    m = (date_yyyymmdd // 100) % 100
    d = date_yyyymmdd % 100
    h = ms_of_day // 3_600_000
    mi = (ms_of_day // 60_000) % 60
    s = (ms_of_day // 1000) % 60
    us = (ms_of_day % 1000) * 1000

    local = datetime(y, m, d, h, mi, s, us, tzinfo=_NY, fold=fold)
    utc_nanos = int(local.timestamp() * 1_000_000_000)

    roundtrip = datetime.fromtimestamp(utc_nanos / 1_000_000_000, tz=_NY)
    if roundtrip.hour != h or roundtrip.day != d:
        raise DecodeError(
            f"Non-existent local time (DST spring-forward): date={date_yyyymmdd} ms={ms_of_day}"
        )

    key = (date_yyyymmdd, fold)
    if key not in _DATE_CACHE:
        _DATE_CACHE[key] = utc_nanos - (h * 3_600_000 + mi * 60_000 + s * 1000) * 1_000_000 - us * 1000

    return utc_nanos


def iso_ts_to_unix_nanos(iso_ts: str, fold: int = 0) -> int:
    """Convert a v3 ISO timestamp ("2024-06-20T09:30:00.334") in ET → UTC nanos.

    Raises DecodeError for DST spring-forward gap times.
    """
    # Format is "YYYY-MM-DDTHH:MM:SS" or "YYYY-MM-DDTHH:MM:SS.fff".
    try:
        if "." in iso_ts:
            naive = datetime.strptime(iso_ts, "%Y-%m-%dT%H:%M:%S.%f")
        else:
            naive = datetime.strptime(iso_ts, "%Y-%m-%dT%H:%M:%S")
    except (TypeError, ValueError) as exc:
        raise DecodeError(f"Invalid ISO timestamp: {iso_ts!r}") from exc

    local = naive.replace(tzinfo=_NY, fold=fold)
    utc_nanos = int(local.timestamp() * 1_000_000_000)

    roundtrip = datetime.fromtimestamp(utc_nanos / 1_000_000_000, tz=_NY)
    if roundtrip.hour != naive.hour or roundtrip.day != naive.day:
        raise DecodeError(f"Non-existent local time (DST spring-forward): {iso_ts!r}")
    return utc_nanos


def _decimal_places(value: float | int) -> int:
    s = str(value)
    if "." in s:
        return len(s.rsplit(".", 1)[1])
    return 0


def _price_from_wire(value, price_precision: int) -> Price:
    """Convert a wire numeric value to a Price at the declared precision.

    Raises DecodeError on non-numeric input or precision drift.
    """
    if not isinstance(value, (int, float)) or isinstance(value, bool):
        raise DecodeError(f"Invalid price (non-numeric): {value!r}")
    n = _decimal_places(value)
    if price_precision < n:
        raise DecodeError(
            f"Price precision mismatch (drops digits): wire={value!r} declared={price_precision}"
        )
    if price_precision not in (n, n + 1):
        raise DecodeError(
            f"Price precision mismatch: wire={value!r} (natural={n}) declared={price_precision}"
        )
    return Price(float(value), price_precision)


def _quantity_from_wire(value, size_precision: int) -> Quantity:
    if not isinstance(value, (int, float)) or isinstance(value, bool):
        raise DecodeError(f"Invalid size (non-numeric): {value!r}")
    if value < 0:
        raise DecodeError(f"Negative size: {value}")
    return Quantity(float(value), size_precision)


def _make_trade_id(iso_ts: str, sequence: int, exchange: int) -> TradeId:
    """Stable TradeId from wire fields. Same inputs → same id."""
    return TradeId(f"{iso_ts}-{sequence}-{exchange}")


# ---------------------------------------------------------------------------
# REST decoders (v3 named-field rows)
# ---------------------------------------------------------------------------


def rest_quote_row_to_quote_tick(
    row: dict,
    instrument_id: InstrumentId,
    price_precision: int,
    size_precision: int,
    ts_init: int,
) -> QuoteTick:
    """Decode one v3 REST historical quote row → QuoteTick."""
    try:
        ts_event = iso_ts_to_unix_nanos(row["timestamp"])
        bid = _price_from_wire(row["bid"], price_precision)
        ask = _price_from_wire(row["ask"], price_precision)
        bid_size = _quantity_from_wire(row["bid_size"], size_precision)
        ask_size = _quantity_from_wire(row["ask_size"], size_precision)
    except KeyError as exc:
        raise DecodeError(f"REST quote row missing field: {exc}") from exc
    return QuoteTick(instrument_id, bid, ask, bid_size, ask_size, ts_event, ts_init)


def rest_trade_row_to_trade_tick(
    row: dict,
    instrument_id: InstrumentId,
    price_precision: int,
    size_precision: int,
    ts_init: int,
) -> TradeTick:
    """Decode one v3 REST historical trade row → TradeTick."""
    try:
        iso = row["timestamp"]
        ts_event = iso_ts_to_unix_nanos(iso)
        price = _price_from_wire(row["price"], price_precision)
        size = _quantity_from_wire(row["size"], size_precision)
        condition = int(row.get("condition", -1))
        sequence = int(row.get("sequence", 0))
        exchange = int(row.get("exchange", 0))
    except KeyError as exc:
        raise DecodeError(f"REST trade row missing field: {exc}") from exc
    if condition == _AGGRESSOR_BUYER:
        aggressor = AggressorSide.BUYER
    elif condition == _AGGRESSOR_SELLER:
        aggressor = AggressorSide.SELLER
    else:
        aggressor = AggressorSide.NO_AGGRESSOR
    trade_id = _make_trade_id(iso, sequence, exchange)
    return TradeTick(instrument_id, price, size, aggressor, trade_id, ts_event, ts_init)


def rest_ohlc_row_to_bar(
    row: dict,
    bar_type: BarType,
    price_precision: int,
    ts_init: int,
) -> Bar:
    """Decode one v3 REST historical OHLC row → Bar. `vwap` is dropped (no Nautilus field)."""
    try:
        ts_event = iso_ts_to_unix_nanos(row["timestamp"])
        # v3 OHLC rows include zero-volume "padding" rows (open=high=low=close=0.0,
        # volume=0). Nautilus's Bar invariants don't accept that — high must be
        # > low, etc., and prices=0 violate the implicit non-negative tick.
        # Caller is responsible for filtering; emit the raw row here.
        open_ = _price_from_wire(row["open"], price_precision)
        high = _price_from_wire(row["high"], price_precision)
        low = _price_from_wire(row["low"], price_precision)
        close = _price_from_wire(row["close"], price_precision)
        volume = _quantity_from_wire(row["volume"], 0)
    except KeyError as exc:
        raise DecodeError(f"REST OHLC row missing field: {exc}") from exc
    return Bar(bar_type, open_, high, low, close, volume, ts_event, ts_init)


# ---------------------------------------------------------------------------
# WebSocket decoders (v3 frame shape)
# ---------------------------------------------------------------------------
#
# WS frame payload schema not fully documented for v3; the live-probe TODO in
# the wire-format note covers verification. These decoders assume the v2 shape
# with `quote.bid` / `trade.price` etc. — adjust here when live capture lands.


def ws_quote_frame_to_quote_tick(
    frame: dict,
    instrument_id: InstrumentId,
    price_precision: int,
    size_precision: int,
    ts_init: int,
) -> QuoteTick:
    """Decode one WS QUOTE frame → QuoteTick."""
    try:
        quote = frame["quote"]
        bid = _price_from_wire(quote["bid"], price_precision)
        ask = _price_from_wire(quote["ask"], price_precision)
        bid_size = _quantity_from_wire(quote["bid_size"], size_precision)
        ask_size = _quantity_from_wire(quote["ask_size"], size_precision)
    except KeyError as exc:
        raise DecodeError(f"WS quote frame missing field: {exc}") from exc
    # v3 WS payloads either carry a `timestamp` ISO string or none at all;
    # fall back to ts_init for live (arrival) timestamp.
    ts_event = ts_init
    iso = quote.get("timestamp")
    if iso:
        try:
            ts_event = iso_ts_to_unix_nanos(iso)
        except DecodeError:
            ts_event = ts_init
    return QuoteTick(instrument_id, bid, ask, bid_size, ask_size, ts_event, ts_init)


def ws_trade_frame_to_trade_tick(
    frame: dict,
    instrument_id: InstrumentId,
    price_precision: int,
    size_precision: int,
    ts_init: int,
) -> TradeTick:
    """Decode one WS TRADE frame → TradeTick."""
    try:
        trade = frame["trade"]
        price = _price_from_wire(trade["price"], price_precision)
        size = _quantity_from_wire(trade["size"], size_precision)
    except KeyError as exc:
        raise DecodeError(f"WS trade frame missing field: {exc}") from exc
    condition = trade.get("condition")
    if condition == _AGGRESSOR_BUYER:
        aggressor = AggressorSide.BUYER
    elif condition == _AGGRESSOR_SELLER:
        aggressor = AggressorSide.SELLER
    else:
        aggressor = AggressorSide.NO_AGGRESSOR
    iso = trade.get("timestamp") or ""
    sequence = int(trade.get("sequence") or 0)
    exchange = int(trade.get("exchange") or 0)
    trade_id = _make_trade_id(iso, sequence, exchange)
    ts_event = ts_init
    if iso:
        try:
            ts_event = iso_ts_to_unix_nanos(iso)
        except DecodeError:
            ts_event = ts_init
    return TradeTick(instrument_id, price, size, aggressor, trade_id, ts_event, ts_init)
